### ISSUE TITLE
allow user to request consistent ordering of grouped alerts

### ISSUE THREAD

**Issue by tzz** [prometheus/alertmanager#1178]
alertmanager currently groups alerts without a consistent ordering, which is tough to follow for large groups.

The user should be able to specify a sort order on some label or annotation, which would give consistency and locally meaningful sorting. This could be done with a Go sort function in the template or with a alertmanager configuration entry perhaps.

**Comment by brian-brazil** [comment]
I propose that at a minimum the `Alerts` in the template structure should come with some ordering on it, like we have done elsewhere.

**Comment by stuartnelson3** [comment]
Are you referring to the `/api/v1/alerts/groups` endpoint? we're currently sorting by internal identifier in `/api/v1/alerts` (https://github.com/prometheus/alertmanager/blob/master/api/api.go#L405-L407)

**Comment by brian-brazil** [comment]
This is about notification templates.

**Comment by tzz** [comment]
I don't mind writing some code if it will help move this forward. For me it would be enough to have a function or alert manager option to simply sort alerts lexicographically by label or annotation.

**Comment by brian-brazil** [comment]
I think we should start with a consistent ordering on the Alerts we provide to notification templates. That way users get something okay without having to do extra work.

**Comment by tzz** [comment]
For a default ordering, how about lexicographical by `(alert.annotations.summary, alert.creation_time)`? That's the simplest thing I can see that could be generally useful.

**Comment by brian-brazil** [comment]
You can't presume that any particular annotation exists, nor that creation times are stable. I'd suggest working entirely off alert labels. This should not be configurable at this level.

**Comment by tzz** [comment]
OK, what alert labels would make for a sensible default sort key? `(job, instance)`?

**Comment by brian-brazil** [comment]
You'll need to use all of them, otherwise it won't be consistent. Moving job and instance to the front is probably wise.

**Comment by tzz** [comment]
Looking at `dispatch/dispatch.go`, there seem to be two ways to go: either modify `aggrGroup` to always have a sorted list of `Alert` structs instead of a map, or sort the `alertsSlice` when alerts are flushed. The `alertsSlice` sort seems easier to implement.

Does that make sense or am I misunderstanding your intention or the code?

**Comment by brian-brazil** [comment]
Yes, the slice would be the one to sort somewhere along that codepath.

**Comment by tzz** [comment]
I'm having trouble compiling the code because of some strange unrelated error `vendor/github.com/weaveworks/mesh/token_bucket.go:33: undefined: time.Until
` so I can't verify the following, and Go is not a strong language for me. So I'm putting my proposed approach here (to be placed in dispatch/dispatch.go:flush():L426). I hope this is consistent with your thinking. If this trivial patch is enough and only needs minor fixes, maybe you or someone else can add it? Or it can wait until the compilation error is resolved.

```
	// look at labels.job, then labels.instance
	sort.SliceStable(alertsSlice, func(i, j int) bool {
		ki := alertsSlice[i].Labels.get("job")
		kj := alertsSlice[j].Labels.get("job")
		if ki != kj {
			return ki > kj
		}

		ki := alertsSlice[i].Labels.get("instance")
		kj := alertsSlice[j].Labels.get("instance")
		if ki != kj {
			return ki > kj
		}

		return String(alertsSlice[i].Labels) > String(alertsSlice[j].Labels)
	})
```

**Comment by tzz** [comment]
The proposed change is in https://github.com/prometheus/alertmanager/pull/1234 but if Alertmanager will allow arbitrary sorting from the user in the future, then the `LabelSet.Before()` method should probably be extended to take a list of label names, in which case the change becomes trivial. I didn't propose that change because I don't know if it's right, and wanted to keep the scope as small as possible.

**Comment by tzz** [comment]
@brian-brazil the request was for the user to be able to request a specific sort order for alerts and I thought #1234 was just the first step. Does closing this request mean it won't be done or that it will be implemented elsewhere?

**Comment by brian-brazil** [comment]
Ah, I'd missed that. I'm personally hoping we can avoid having to implement that.

**Comment by tzz** [comment]
Another approach would be to sort by a special `sort_key` annotation first, then by `job` and `instance`. It wouldn't be hard to support that as an extension of #1234. It would solve all my needs at least.

**Comment by aclowkey** [comment]
I'm curious about the way the API returns the result. is there any sorting logic?

**Comment by OlafKocanda** [comment]
I have a suggestion:
Couldn't you just order in the sequence of the group by list?

e.g. 
group_by: ['severity', 'alertname'] 
leads to first grouping by severity and then by alertnames

group_by: ['alertname','severity'] 
first groups by alertnames then by severity

**Comment by hwreschnig-utiq** [comment]
We have a use case for this where we're generating metrics per domain, writing alerts to detect various kinds of traffic anomalies, then grouping notifications "per-anomaly" to produce a list of domains with each type of anomaly. This can be dozens of items and we would prefer to get this list of domains sorted when formatting it.

My proposal would be e.g. for our case,
```
    group_sort_key_labels: [domain]
```

and if none is set, skipping the sort to keep whatever the current order would be. Notably, we *don't* need dynamic sorting, multiple sorts, or to do any kind of sorting *in* the template, we just want the alerts slice sorted in a particular order *for* the template.

If there's no resistance to this approach I would like to prepare a PR for it.
